src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala

```diff
@@ -553,30 +555,45 @@ trait TypeDiagnostics extends splain.SplainDiagnostics {
 
     def handleRefTree(t: RefTree): Unit = {
       val sym = t.symbol
-      if (isExisting(sym) && !currentOwner.hasTransOwner(sym) && !t.hasAttachment[ForAttachment.type])
+      if (isExisting(sym) && !isCurrentlyEnclosedBy(sym) && !t.hasAttachment[ForAttachment.type])
         recordReference(sym)
     }
 
+    // sym is an owner or an owner is a synthetic case method owned by sym's companion module
+    // note: isConstructor avoids only super.<init> at typer
+    def isCurrentlyEnclosedBy(sym: Symbol): Boolean =
+      currentOwner.hasTransOwner(sym) || sym.isCaseClass && {
+        val module = sym.companionModule.moduleClass
+        def enclosed =
+          currentOwner.ownersIterator
+            .takeWhile(owner => !owner.isClass && !owner.isModuleClass)
```

Comment:
```suggestion
            .takeWhile(owner => !owner.isClass)
```

module classes are classes